### ISSUE TITLE
Games don't work under Proton 5.13-6 when run from Steam client

### ISSUE THREAD

**Issue by PookyFan** [ValveSoftware/steam-runtime#390]
**System info**:
https://gist.github.com/PookyFan/2de98d30ecb28047235c29e6ac3de692

**Other information**:
Linux Distro: Debian 10
Linux kernel: 5.10.26 with MuQSS scheduler
GPU driver version: 460.39

**Problem description**:
For a reason unknown, no game runs under Proton 5.13-6 when launched from Steam client. After launching some games (ie _Senran Kagura Burst Re:Newal_ or _Rabi-Ribi_), the Play button returns from Playing state back to the original in a few moments.
Some other games (ie first _Dawn of War_) would just hang on launching, with no window appearing and Playing button not returning to Play until cancelling game launch.

When I use Proton 5.0-10, games launch normally.

The logs mention a few errors, although I'm not sure which one is the culprit.
_Senran Kagura_:
[steam-889510.log](https://github.com/ValveSoftware/Proton/files/6245299/steam-889510.log)

_Rabi-Ribi_:
[steam-400910.log](https://github.com/ValveSoftware/Proton/files/6245462/steam-400910.log)

_Dawn of War_:
[steam-4570.log](https://github.com/ValveSoftware/Proton/files/6245354/steam-4570.log)

Interestingly, when I use Proton to run non-Steam game (for example _Genshin Impact_) from the console like so:
`export STEAM_COMPAT_DATA_PATH=/home/pookyfan/gaemz/steamapps/compatdata
"home/pookyfan/gaemz/steamapps/common/Proton 5.13/proton" run launcher.exe`
then this game works perfectly fine. Steam games can't be launched this way, however - it ends with failure for most of tested games (with _VRChat_ being an exception - I was able to run it from console, but I couldn't login with Steam account, obviously).

**Comment by RyuzakiKK** [comment]
@PookyFan Your "System Information" gist tells us that the Soldier runtime failed to setup correctly (the last `CClientJobGatherRuntimeInformation: container read output failed` line in your report).

To investigate the cause of that you should post your `slr-app*-*.log` and/or `slr-non-steam-game-*.log` file that you can obtain by following this guide https://github.com/ValveSoftware/steam-runtime/blob/master/doc/reporting-steamlinuxruntime-bugs.md#essential-information

Before gathering the required logs, be sure to revert all your manual modifications (you can do it on the Steam client by right clicking on Steam Linux Runtime - Soldier and then Properties->Local Files->Verify integrity of tool files)

**Comment by RyuzakiKK** [comment]
@jtrees Without any further info, for now let's treat your issue as a separate one. Please open a new issue report, including all the required info from the Github template, specifically the Steam->Help->System Information and the Soldier log while attempting to open the game that doesn't work https://github.com/ValveSoftware/steam-runtime/blob/master/doc/reporting-steamlinuxruntime-bugs.md

**Comment by PookyFan** [comment]
@RyuzakiKK Thanks for the insights. I'm attaching requested logs from launching _Senran Kagura_ with Proton 5.13-6:
[slr-app889510-se76fb3cb3ebdb1eb.log](https://github.com/ValveSoftware/steam-runtime/files/6258422/slr-app889510-se76fb3cb3ebdb1eb.log)

**Comment by RyuzakiKK** [comment]
@PookyFan the only problem that I can spot from your log is the one related to the locale generation, where `pressure-vessel-locale-gen` signals that it created some locales that were missing, but the directory where the locales should be generated is empty.

Can you please run Steam with `STEAM_LINUX_RUNTIME_LOG=1 STEAM_LINUX_RUNTIME_VERBOSE=1 steam`  and try again to launch the game and post the log? This will create a log that is more verbose and should include the diagnostic output of `pressure-vessel-locale-gen` that hopefully contains some valuable information about what is causing this issue.

**Comment by PookyFan** [comment]
@RyuzakiKK sure. Attaching verbose log:
[slr-app889510-s9b9462f867ae26b9.log](https://github.com/ValveSoftware/steam-runtime/files/6258582/slr-app889510-s9b9462f867ae26b9.log)

**Comment by RyuzakiKK** [comment]
```
pressure-vessel-locale-gen: Missing locale en_US.UTF-8
pressure-vessel-locale-gen: Generating locale en_US.UTF-8...
/run/pressure-vessel/pv-from-host/bin/pressure-vessel-locale-gen: 182: /run/pressure-vessel/pv-from-host/bin/pressure-vessel-locale-gen: localedef: not found
pressure-vessel-locale-gen: Unable to generate locale en_US.UTF-8: 127
setlocale "en_US.UTF-8": No such file or directory
```

And it seems like we bind `localedef` inside the container from the host's `/usr/local/bin/localedef`, instead of the canonical `/usr/bin/localedef` provided by Debian's `libc-bin`.

I assume that `/usr/local/bin/localedef` is something that you compiled yourself? Can you try to remove/rename `/usr/local/bin/localedef` and `/usr/local/bin/locale`?

**Comment by PookyFan** [comment]
@RyuzakiKK yeah, I have a custom build of GCC toolchain with C and C++ libraries placed in `/usr/local` directory. Never thought it would cause issues with Steam.

I'd prefer to keep my custom toolchain and libraries this way and make Steam work with it, but I assume it may not be supported, and if that's the case, I guess this issue should be closed. Still, can you please look once again at the logs after I renamed both files as you requested? It looks like this didn't solve the problem, even though locale generation issue seems to be gone. If that's something that may be taken care of easily, I'd like to do that and make newer Proton versions work correctly with the runtime. If not, then - as I said - there is no need to keep this issue open.

Non-verbose log:
[slr-app889510-s1dfb99e9b4ec44e0.log](https://github.com/ValveSoftware/steam-runtime/files/6258833/slr-app889510-s1dfb99e9b4ec44e0.log)

Verbose log:
[slr-app889510-s40ca5c29e6ebaa.log](https://github.com/ValveSoftware/steam-runtime/files/6258846/slr-app889510-s40ca5c29e6ebaa.log)

**Comment by smcv** [comment]
> I have a custom build of GCC toolchain with C and C++ libraries placed in /usr/local directory. Never thought it would cause issues with Steam

If it's in search paths like `PATH`, then it's going to affect anything and everything - you're effectively acting as your own OS vendor, and making your own Debian-derived operating system with different C and C++ runtime libraries.

If you don't intend these to be used by typical programs and only intend them to be used on-demand, then I'd recommend installing in a separate `${prefix}`, perhaps `/opt/newgcc` or something.

**Comment by smcv** [comment]
I'm guessing your locally-installed `localedef` has a non-standard `DT_RUNPATH`, `DT_RPATH` or ELF interpreter, which exists on the host system but not inside the container?

You seem to be using glibc 2.28 (which is normal for Debian 10) for 32-bit programs, but glibc 2.32 (which is newer than Debian 11!) for 64-bit programs. There's a limit to how well we can support systems where something as fundamental as the Standard C runtime library has been changed.

**Comment by PookyFan** [comment]
I understand that my customization of the OS libs may really be at fault. But as I said, my setup was done this way deliberately and I don't want to give up on it. Guess I need to either stick to Proton 5.0 or keep Soldier runtime disabled, as this workaround seems to be working fine with newer Proton versions. For now I'll close this issue and maybe come back to it when / if I decide to install another linux distro with no such modifications and I happen to experience similar problems.

Thanks for all insights and comments, it's nice you guys spent your time analyzing this.

**Comment by smcv** [comment]
> or keep Soldier runtime disabled

I would not recommend this, and we cannot support it. Expect it to randomly stop working at the least convenient time.

> I'm guessing your locally-installed localedef has a non-standard DT_RUNPATH, DT_RPATH or ELF interpreter, which exists on the host system but not inside the container?

If you want us to be able to improve pressure-vessel (the container tool used for soldier) to cope with your setup, then we are going to need more information about how things like this line up.

**Comment by PookyFan** [comment]
> I would not recommend this, and we cannot support it. Expect it to randomly stop working at the least convenient time.

I'll keep that in mind, thanks. Worst case scenario is sticking to Proton 5.0, for now all games I'm playing work fine with it anyway.

> If you want us to be able to improve pressure-vessel (the container tool used for soldier) to cope with your setup, then we are going to need more information about how things like this line up.

Well, it would be nice if it worked with my setup, but I'm not sure what kind of information we're talking about. Is it about how exactly I customized my libs, or...?

In my case, I have built myself GCC 10.2.0 with bintutils 2.36.1 and glibc 2.32 - all placed in `/usr/local` to take priority over default tools and libs installed in OS and managed by apt. Stuff like ld and such was also included (although it's accessible from the default `/lib64/ld-linux-x86-64.so.2` symlink), so yes, there is new ELF interpreter in my OS now, and it's probably different from one from the (Steam related, I assume?) container. Not much beyond that, I guess. As I didn't need to build custom 32 bit libs, I just fetched them from apt, as they would end up being used by Steam and its games only.
